src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -177,6 +177,43 @@ func logHostsStatus(log logrus.FieldLogger, hosts map[string]inventory_client.Ho
 	log.Infof("Hosts status: %v", hostsStatus)
 }
 
+func (c *controller) updateNodesLabels() bool {
+	ignoreStatuses := []string{models.HostStatusDisabled}
+	ctxReq := utils.GenerateRequestContext()
+	log := utils.RequestIDLogger(ctxReq, c.log)
+
+	assistedNodesMap, err := c.ic.GetHosts(ctxReq, log, ignoreStatuses)
+	if err != nil {
+		log.WithError(err).Error("Failed to get node map from the assisted service")
+		return KeepWaiting
+	}
+
+	nodes, err := c.kc.ListNodes()
+	if err != nil {
+		log.WithError(err).Error("Failed to get list of nodes from k8s client")
+		return KeepWaiting
+	}
+
+	for _, node := range nodes.Items {
+		hostData, ok := assistedNodesMap[strings.ToLower(node.Name)]
+		if !ok {
+			log.Warnf("Node %s is not in inventory hosts", strings.ToLower(node.Name))
+			continue
+		}
+
+		if len(hostData.Host.NodeLabels) == 0 {
+			continue
+		}
+
+		err = c.kc.PatchNodeLabels(&node, hostData.Host.NodeLabels)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to patch %s node with labels %s", node.Name, hostData.Host.NodeLabels)
+			continue
+		}
+	}
+	return ExitWaiting
```

Comment:
How do you know you've successfully labeled all the nodes that are going to join?

Some nodes join much later than other nodes.

e.g. workers join after control plane nodes - in that case, "ListNodes" at first would give you just the control plane nodes, and only later once the workers join you'll see the worker nodes in that list.

I think your method should loop over `assistedNodesMap`, not over `nodes`, and only once it successfully patched all of them, it should ExitWaiting.

You should also stop ignoring errors - you should keep trying until you've successfully patched all the nodes with the right labels.

Reply:
@omertuc This func is under PostInstallConfigs which is after WaitAndUpdateNodesStatus. I might be mistaken here but according to my understanding, all the nodes would have already joined by the time we get to PostInstallConfigs. So this should not be an issue. Thoughts?

Reply:
The issue is, is that the finalizing state doesn't imply that all nodes have already joined. Here is why:

The transition between installing and finalizing:
https://github.com/openshift/assisted-service/blob/e251185323c16862f4898380895e0954e86180b0/internal/cluster/statemachine.go#L254-L264

Is fired when the IsFinalizing method is true:

https://github.com/openshift/assisted-service/blob/b365b892b2288e4c4914224e3e3a18db7e84f24f/internal/cluster/transition.go#L328-L339

Reply:
Also don't forget this part of my comment, it's also important:

>You should also stop ignoring errors (i.e. continue) - you should keep trying until you've successfully patched all the nodes with the right labels.

Reply:
Understood. And yes, I have taken into account the "stop ignoring errors" part that you commented.